### DISCUSSION TITLE
주요 도메인

### DISCUSSION THREAD

**Discussion by Bue-von-hon** [f-lab-edu/foody-lab#8] [Ideas]
주문, 결제, 주문 처리 도메인 3가지의 프로세스에 대해 자유롭게 말씀해주세요
주요 프로세스는 [여기서](https://github.com/f-lab-edu/foody-lab/wiki/foody-lab%EC%9D%98-%EA%B8%B0%EB%8A%A5) 확인 가능합니다.

**Comment by Taewoongjung** [comment]
주문 도메인 안에 “주문 생성“, “주문 처리” 가 소주제로 있어야 하는게 아닐까요?

**Reply by Bue-von-hon** [reply]
저도 사실 해당 부분이 헷갈렸습니다!
왜냐하면 주문의 생성과 처리가 한번에 묶이는게 자연스러워 보이니까요
하지만 주문 생성은 "고객"이, 주문 처리는 "점주"가 진행하고 있거든요
따라서 주문을 "생성"하는 영역과 주문을 "처리"하는 영역은 별개라고 생각이 들었어요
p.s. 일단 두개를 묶어서 진행해볼까요? 아무래도 이를 진행시키는 주체가 달라서 저도 고민입니다...ㅠㅠ

**Reply by Taewoongjung** [reply]
네 묶어서 진행해보아요! 

<img width="627" alt="image" src="https://user-images.githubusercontent.com/70272679/194230534-f71adf1a-41a4-4a45-ae2d-83dbb263c431.png">

저는 주문 "생성"과 "처리"는 이런 느낌이라 보거든요.

**Reply by Bue-von-hon** [reply]
넵 그럼 바로 위키를 수정하겠습니다!

**Reply by Taewoongjung** [reply]
그리고 도메인을 정할때 "_라이프 사이클"_ 을 따져보면 도메인 나누기 쉬울 것 같네요, 
주문을 넣을 때 주문객체를 생성하려면 가게 객체도 생성되어야 하는데 가게 객체를 생성 하려면 가게에 해당하는 점주도 생성해야겠네요. 그렇게 되면 같은 도메인이라고 생각할 수 있겠습니다.

**Reply by Bue-von-hon** [reply]
주문과 가게, 가게 오너까지 묶을필요는 없을것 같아요
주문과 가게는 함께 생성되는것이 아니라, 따로 생성될것 같거든요!

ex) 고객이 주문할때마다 가게나 가게 오너가 생성되는건 자연스럽지 않은것 같아요!

**Reply by Taewoongjung** [reply]
네 계속 같이 생성한다는 의미는 아니었고 따로 미리 생성이 되어야 한다는 의미었는데 같이 생성한다로 들릴 수 있었겠네요! 죄송합니다 ~

**Comment by Taewoongjung** [comment]
주문 도메인에 있는 "주문 생성시 어떤 결제 타입을 사용할건지 결정해야만 한다."는 **결제 도메인**에 어울릴 것 같아요. 

<img width="899" alt="image" src="https://user-images.githubusercontent.com/70272679/194231513-94e62c77-23b5-47ce-9c8c-1bee2a061ebe.png">

이런 느낌으로 **어떤 결제 타입을 사용할건지 결정**하는 것은 주문 도메인에서 주문 정보를 생성할 때 넣어야할 정보인 것 같아보이네요.

**Reply by Bue-von-hon** [reply]
현재도 주문 정보를 생성할때 결제타입을 결정하도록 프로세스를 정의하고 있어서요, 잘 적어놨다는 말씀이시죠??

**Reply by Taewoongjung** [reply]
아하 어떤 카드로 결제를 할건지랑 카드결제, 계좌이체, 현장결제 같은 종류들이랑 헷갈렸네요. 어떤 카드로 결제할건지 같은 경우만 아니면 주문 도메인에 있어도 된다고 생각합니다!

**Comment by Taewoongjung** [comment]


**Reply by Bue-von-hon** [reply]
언급해주신 내용은 주문 생성 제약 조건으로 생각되거든요
주문 생성 제약 조건이 결제 도메인으로 이동해야하는 이유가 있을까요..?

**Reply by Taewoongjung** [reply]
> 언급해주신 내용은 주문 생성 제약 조건으로 생각되거든요 주문 생성 제약 조건이 결제 도메인으로 이동해야하는 이유가 있을까요..?

DM으로 답 완료.
